### DISCUSSION TITLE
How to enable SSR support for 'styled-components' for third party library(NextJS+styled-components+babel+component library)

### DISCUSSION THREAD

**Discussion by palakurthivishal** [babel/babel#13434] [Q&A]
styled components for SSR needs to be transpiled with special plugins such as (babel-plugin-styled-components or typescript-plugin-styled-components)

https://styled-components.com/docs/tooling#serverside-rendering

Using so, it adds the deterministic component ID to each of the components to avoid classname collision and thus to avoid hydration issues such as Classname didn't match

my babel.config.json in NextJS app is set to,

```
{
  "presets": ["next/babel"],
  "plugins": [["styled-components", { "ssr": true, "displayName": false, "minify": true }]]
}
```
this takes care of enabling SSR support for the styled components which I have created in the app(./src/components) BUT

I'm consuming a third-party component library(@org/carousel) which is not maintained by me so I have no control over its build & publish process.

Since NextJS clearly ignores transpiling node_modules, so I have looked at next-transpiled-modules with the following setup in next.config.js, hoping that it would that support

```
const withTM = require('next-transpile-modules')(['@org/carousel']);

module.exports = withTM();
```

Yet, in my build, I don't see the addition of deterministic ID for that component.

Gist: node_modules/@org/carousel/build/index.js has styled.div() but we need it as styled.div.withConfig({componentId:'sc-someId'})

- Am I in the right direction?
- Can we transpile an ES5 compliant module again? (@org/carousel was already transpiled before publish)
- Are there any other ways to address this issue?

Any thoughts/ideas/suggestions are greatly appreciated!

**Comment by nicolo-ribaudo** [comment]
You can try running `BABEL_SHOW_CONFIG_FOR="./node_modules/@org/carousel/build/index.js" npm run build` (using your build command instead of `npm run build`) to see what config is being applied to that file.

> Can we transpile an ES5 compliant module again? (@org/carousel was already transpiled before publish)

Yes and no. Technically yes, however this only works if the plugin doesn't make particular assumptions about the code it's compiling.

For example, if JSX is already compiled out, a plugin that expects to run on JSX code won't work. `babel-plugin-styled-components` has some [complex logic](https://github.com/styled-components/babel-plugin-styled-components/blob/950692b92855f17609e4281d5081e3d2acbf4b6b/src/visitors/displayNameAndId.js#L143) to check if it's compiling a `styled.div()` call, and if in the compiled code is not exactly `styled.div()` it might not detect it.

> Are there any other ways to address this issue?

If the problem is what I mentioned above, you could try compiling the _untranspiled_ source of `@org/carousel`.

**Reply by palakurthivishal** [reply]
Thanks for the response @nicolo-ribaudo , you have pointed to the right source 👍  , I went through the plugin's implementation by taking a glimpse of AST at first.

**This plugin expects "styled-components" to be an import/require in the file** --> in my case, @org/carousel didn't import that way :( , instead they have bundled "styled" as a named export in one of its dependencies and made use of it. Like below,
```
 // @org/theme
 import styled from "styled-components";
 
 const theme = {};
 export {styled};
 export default theme;
 
 // @org/carousel
 import {styled} from '@org/theme'; // --> doesn't comply with babel-plugin-styled-component
 var Wrapper = styled('div') ...
```

Due to this, I have adjusted the plugin's code to detect my use case correctly (I fiddled with node_modules/babel-plugin-styled-components) and has successfully managed to transpile @org/carousel **again** (I have verified the build, sc now has **withConfig** )


**But**, the class name collision is still happening :(, I still get **Class name doesn't match** error in my NextJS app. The class names on the server and the client aren't the same due to which React is failing to hydrate and my app styles are broken.

To debug further, I have enabled the **displayName** option on the plugin. To my surprise, the className generated on the server doesn't have the displayName prefix but the one on the client-side has it. This is wrong behavior but I don't yet know the reason.

I'm currently at two junctions,
- **@org/carousel didn't handle class name as suggested [here](https://styled-components.com/docs/advanced#caveat)**
- **I have seen this in one of our components, not sure if this is legit but this was working for us until a couple of days ago.**

```
 //file1.js
 import styled from 'styled-components";
 
 export const Wrapper = styled.div``;
 export const Heading = styled.h1`color:red`;
 export default function Comp1(props){
   return <Wrapper {...props}><Heading>Hello</Heading></Wrapper>;
 }
 
 //file2.js
  import styled from 'styled-components";
  import Comp1, { Heading} from './file1';
  
  const AnotherWrapper = styled.div`
   ${Heading} {
    color: blue; // ** is this type of overriding correct? this way works fine in CSR, how about SSR?**
   }
 `
  export default function Comp2() {
    return <AnotherWrapper><Comp1>World..!</AnotherWrapper>;
  }
```

I'm highly determined to understand the cause of this issue and resolve it.
